### ISSUE TITLE
Translation quality does not improve

### ISSUE THREAD

**Issue by ben15021999** [facebookresearch/fairseq#4282]
#### What is your question?
I has implemented reward shaping RL method in criterion. I was to use SequenceGenerator to translate sentences and calculate GLEU as a reward. It seems very fast and make lower loss but when I use fairseq-generate, the quality of translation does not change at all. I don't know why, please help!

#### Code
        tgt_dict = self.task.target_dictionary
        eos_idx = self.task.target_dictionary.eos()
        sample_beam = self.beam_size
        search_strategy = (
            search.Sampling(tgt_dict, sampling_topk=self.sampling_topk) if self.multinomial_sample_train else None
        )
        translator = SequenceGenerator([model], tgt_dict=tgt_dict,
                                       beam_size=sample_beam, min_len=1, search_strategy=search_strategy)
        translator = utils.move_to_cuda(translator)
        ct = 0
        translations = []

        s = utils.move_to_cuda(sample)
        input = s['net_input']
        hypos = translator.generate(
            model,
            sample,
        )
        for i, id in enumerate(s['id'].data):
            src = input['src_tokens'].data[i, :]
            # remove padding from ref
            ref = utils.strip_pad(s['target'].data[i, :], tgt_dict.pad()) if s['target'] is not None else None
            translations.append((id, src, ref, hypos[i]))
            ct += 1
        model.train(True)

        # RL loss
        batch_rl_loss = 0
        batch_tokens = 0
        sample_ind = 0
        for sample_id, src_tokens, tgt_tokens, hypos in translations:
            # calculate bleu
            hypo = hypos[0]
            trans_tokens = hypo['tokens']
            rewards = torch.Tensor(len(trans_tokens)).float().cuda()
            scores = utils.move_to_cuda(-hypo['positional_scores'])
            for i in range(len(trans_tokens)):
                trans = trans_tokens[:(i+1)]
                rewards[i] = self.compute_gleu(tgt_tokens.cpu(), trans.cpu(), max_order=self.max_order, gram=self.gram).cuda()
            rewards = rewards + torch.sum(rewards, dim=0, keepdims=True) - torch.cumsum(rewards, dim=0) # cumulative reward
            rl_loss = torch.sum(scores * rewards)
            batch_rl_loss += rl_loss
        batch_rl_loss = batch_rl_loss.clone().detach().requires_grad_(True)
        avg_rl_loss = batch_rl_loss
        total_tokens = sample['target'].size(0)

        logging_output = {
            'loss': utils.item(batch_rl_loss.data),
            'ntokens': total_tokens,
            'nsentences': total_tokens,
            'sample_size': total_tokens,
        }
        return avg_rl_loss, total_tokens, logging_output
